### PR TITLE
[core] Fix bugs in data locality

### LINKED ISSUE #24267: [Datasets][core] Data locality not working as expected during map stages
### What happened + What you expected to happen

In cases where all inputs are already in Ray's object store, it should be possible to execute a map stage with little to no data movement. Datasets currently does this by relying on Ray core's data locality.

However, it doesn't seem to be working as expected. In the following screenshot of a synthetic sort benchmark, you can see that there is significant network usage between 12:45-12:50, which corresponds to transferring task args during the sample and map phase. This is not possible to eliminate with fusion only, since we need to sample all blocks before executing the map stage.

Screenshot from 2022-04-27 12-07-28

### Versions / Dependencies

2.0dev

### Reproduction script

`python release/nightly_tests/dataset/sort.py ...`

Must run on a cluster.

### Issue Severity

_No response_

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This fixes two bugs in data locality:
1. When a dependent task is already in the CoreWorker's queue, we ran the data locality policy to choose a raylet before we added the first location for the dependency, so it would appear as if the dependency was not available anywhere.
2. The locality policy did not take into account spilled locations.

Added C++ unit tests and Python tests for the above.

## Related issue number

Fixes #24267.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by jjyao — APPROVED** [review]

**Review by mwtian — APPROVED** [review]

File: src/ray/core_worker/task_manager.cc [hunk 1]
```diff
@@ -243,11 +243,14 @@ void TaskManager::CompletePendingTask(const TaskID &task_id,
     const auto nested_refs =
         VectorFromProtobuf<rpc::ObjectReference>(return_object.nested_inlined_refs());
     if (return_object.in_plasma()) {
+      // NOTE(swang): We need to add the location of the object before marking
+      // it as local in the in-memory store so that the data locality policy
+      // will choose the right raylet for any queued dependent tasks.
+      const auto pinned_at_raylet_id = NodeID::FromBinary(worker_addr.raylet_id());
+      reference_counter_->UpdateObjectPinnedAtRaylet(object_id, pinned_at_raylet_id);
```
Comment (clarkzinzow) [inline src/ray/core_worker/task_manager.cc:250]: Ah nice ordering catch!

File: src/ray/core_worker/reference_count.cc [hunk 2]
```diff
@@ -1281,7 +1281,10 @@ absl::optional<LocalityData> ReferenceCounter::GetLocalityData(
   //   locations.
   // - If we don't own this object, this will contain a snapshot of the object locations
   //   at future resolution time.
-  const auto &node_ids = it->second.locations;
+  auto node_ids = it->second.locations;
+  if (!it->second.spilled_node_id.IsNil()) {
+    node_ids.emplace(it->second.spilled_node_id);
+  }
```
Comment (clarkzinzow) [inline src/ray/core_worker/reference_count.cc:1287]: Nice!

File: python/ray/tests/test_scheduling_2.py [hunk 3]
```diff
@@ -647,6 +647,50 @@ def check_backlog_info():
...
+    # the upstream task finishes.
+    for _ in range(5):
+        ray.get(check_locality.remote(f.remote()))
+
+    # Check locality works when some objects were spilled.
+    xs = [f.remote() for _ in range(5)]
+    ray.wait(xs, num_returns=len(xs), fetch_local=False)
+    for i, x in enumerate(xs):
+        task = check_locality.remote(x)
+        print(i, x, task)
```
Comment (clarkzinzow) [inline python/ray/tests/test_scheduling_2.py:690]: Nit: Should remove debug print.

**Review by clarkzinzow — APPROVED** [review]
LGTM if tests pass and nightly test looks good

**Comment by stephanie-wang** [comment]
Still seeing unexpected transfers when there is heavy task load, but this does seem to have fixed the problem in 10TB sort where the head node raylet was coming under heavy load.

Before PR:
Screenshot from 2022-05-09 15-33-28

After PR:
Screenshot from 2022-05-12 23-14-38

Note task throughput (green lline) is very low towards the righthand side of the graph in the before version.

**Comment by stephanie-wang** [comment]
Hmm possibly, but it might just be that the test suite is too large, @mwtian?

**Comment by mwtian** [comment]
That is possible. I guess the test needs to be splitted.

**Comment by mwtian** [comment]
Or is it possible that reconstruction becomes slower? It seems timeouts of 900s happen at or after `test_reconstruction_stress` or `test_reconstruction_hangs`. I will revert this for now..

**Comment by stephanie-wang** [comment]
OK, I'll try to split test_reconstruction after the revert is merged.

**Comment by mwtian** [comment]
Sounds good. If you verified that the slowdown is harmless, I can split the tests too.

**Pull request merged**